lib/lf/fe/hierarchic_fe.h

```diff
@@ -0,0 +1,1227 @@
+          const SCALAR leg3eval = LegendrePoly<SCALAR>::eval(j + 1, l313n);
+          result(3 * degree_ - 1 - j, 2 * i + 0) =
+              l3p1_dx * (j + 2) * std::pow(l3p1, j + 1) * leg3inte +
+              std::pow(l3p1, j + 2) * l313n_dx * leg3eval;
+          result(3 * degree_ - 1 - j, 2 * i + 1) =
+              l3p1_dy * (j + 2) * std::pow(l3p1, j + 1) * leg3inte +
+              std::pow(l3p1, j + 2) * l313n_dy * leg3eval;
+        }
+      }
+      // Get the gradient of the basis functions associated with the interior of
+      // the triangle
+      if (degree_ > 2) {
+        unsigned idx = 3 * degree_;
+        for (unsigned j = 0; j < degree_ - 2; ++j) {
+          SCALAR edge_eval;
+          SCALAR edge_dx;
+          SCALAR edge_dy;
+          if (rel_orient_[1] == lf::mesh::Orientation::positive) {
+            edge_eval = std::pow(l2p3, j + 2) *
+                        LegendrePoly<SCALAR>::integral(j + 2, l233n);
+            edge_dx = result(2 + degree_ + j, 2 * i + 0);
+            edge_dy = result(2 + degree_ + j, 2 * i + 1);
+          } else {
+            edge_eval = std::pow(l2p3, j + 2) *
+                        LegendrePoly<SCALAR>::integral(j + 2, l232n);
+            edge_dx = result(2 * degree_ - j, 2 * i + 0);
+            edge_dy = result(2 * degree_ - j, 2 * i + 1);
+          }
+          for (unsigned k = 0; k < degree_ - j - 2; ++k) {
+            SCALAR jackinte =
+                JacobiPoly<SCALAR>::integral(k + 1, 2 * j + 4, l1[i]);
+            SCALAR jackeval = JacobiPoly<SCALAR>::eval(k, 2 * j + 4, l1[i]);
+            result(idx, 2 * i + 0) =
+                jackinte * edge_dx + edge_eval * jackeval * l1_dx[i];
+            result(idx, 2 * i + 1) =
+                jackinte * edge_dy + edge_eval * jackeval * l1_dy[i];
+            ++idx;
+          }
+        }
+      }
+    }
+    return result;
+  }
+
+  /**
+   * @brief Evaluation nodes are the vertices, the Chebyshev nodes of degree p-1
+   * on the edges and the corresponding nodes on the triangle
+   * @copydoc ScalarReferenceFiniteElement::EvaluationNodes()
+   */
+  [[nodiscard]] Eigen::MatrixXd EvaluationNodes() const override {
+    return eval_nodes_;
+  }
+
+  /**
+   * @brief (p+1)*(p+2)/2 evaluation nodes
+   * @copydoc ScalarReferenceFiniteElement::NumEvaluationNodes()
+   */
+  [[nodiscard]] lf::base::size_type NumEvaluationNodes() const override {
+    return NumRefShapeFunctions();
+  }
+
+  [[nodiscard]] Eigen::Matrix<SCALAR, 1, Eigen::Dynamic> NodalValuesToDofs(
+      const Eigen::Matrix<SCALAR, 1, Eigen::Dynamic> &nodevals) const override {
+    // Simply solfe the LSE for the basis function coefficients
+    const Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic>
+        shape_functions_at_nodes =
+            EvalReferenceShapeFunctions(EvaluationNodes());
+    return shape_functions_at_nodes.transpose()
+        .fullPivHouseholderQr()
+        .solve(nodevals.transpose())
+        .transpose();
+  }
```

Comment:
I suspect that these evaluation nodes lead to a very ill-conditioned system of equations (which needs to be solved in `NodalValuesToDofs`) if p is large...? (Same should be true for segment and quadrilateral)

Reply:
For the segment one can construct linear functionals that should have much better properties using the orthogonality properties of the integrated legendre polynomials. I've derived them just as an example here:

![grafik](https://user-images.githubusercontent.com/15138672/84579550-e8462200-adce-11ea-994f-279cbd9802f2.png)


I think the same procedure can be used to derive linear functionals for the quadrilateral (tensor product structure). For the triangle, it should also work out since the [original paper](https://arxiv.org/pdf/1504.03025.pdf) seems to use the correct jacobi polynomials so that a coordinate transformation can be done.

Reply:
Yes, for the quadrilaterals it's pretty easy to write the dual basis based on the dual basis for segments. The triangles look a bit annoying, but I will try to construct something.

Reply:
If you have problems with the triangles let me know. You probably have to do a variable substitution. I should finally have a bit of time available on the weekend (the last weeks were super busy for me :( )

Reply:
I managed to derive a dual basis for the triangles. However, it looks a bit ugly. I have sent you the PDF over Slack. Maybe you find some trick to simplify it more, but otherwise I will implement it in its current form.